File: impacted-test-engine/src/test/java/com/teamscale/test_impacted/test_descriptor/CucumberPickleDescriptorResolverTest.java
package com.teamscale.test_impacted.test_descriptor;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.LinkedHashMap;

class CucumberPickleDescriptorResolverTest {

	@Test
	void escapeSlashes() {
		LinkedHashMap<String, String> expectedByInput = new LinkedHashMap<>();
		expectedByInput.put("abc", "abc");
		expectedByInput.put("ab/c", "ab\\/c");
		expectedByInput.put("ab\\/c", "ab\\/c"); // don't escape what is already escaped
		expectedByInput.put("/abc", "\\/abc");
		expectedByInput.put("/abc/", "\\/abc\\/");
		expectedByInput.put("/", "\\/");
		expectedByInput.put("/a/", "\\/a\\/");
		expectedByInput.put("a//", "a\\/\\/");
		expectedByInput.put("//a", "\\/\\/a");
		expectedByInput.put("//", "\\/\\/");
		expectedByInput.put("\\", "\\");
		expectedByInput.put("http://link", "http:\\/\\/link");

		expectedByInput.forEach((input, expected) -> Assertions.assertEquals(expected,
				CucumberPickleDescriptorResolver.escapeSlashes(input)));
	}
}

File: impacted-test-engine/src/test/kotlin/com/teamscale/test_impacted/test_descriptor/CucumberPickleDescriptorResolverTest.kt
package com.teamscale.test_impacted.test_descriptor

import com.teamscale.test_impacted.test_descriptor.CucumberPickleDescriptorResolver.Companion.escapeSlashes
import com.teamscale.test_impacted.test_descriptor.CucumberPickleDescriptorResolver.Companion.removeDuplicatedSlashes
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class CucumberPickleDescriptorResolverTest {
	@Test
	fun escapeSlashes() {
		mapOf(
			"abc" to "abc",
			"ab/c" to "ab\\/c",
			"ab\\/c" to "ab\\/c", // don't escape what is already escaped
			"/abc" to "\\/abc",
			"/abc/" to "\\/abc\\/",
			"/" to "\\/",
			"/a/" to "\\/a\\/",
			"a//" to "a\\/\\/",
			"//a" to "\\/\\/a",
			"//" to "\\/\\/",
			"///" to "\\/\\/\\/",
			"\\" to "\\",
			"http://link" to "http:\\/\\/link"
		).forEach { (input, expected) ->
			Assertions.assertEquals(expected, input.escapeSlashes())
		}
	}

	@Test
	fun testNoDuplicatedSlashesInUniformPath() {
		mapOf(
			"abc" to "abc",
			"ab/c" to "ab/c",
			"ab//c" to "ab/c",
			"ab///c" to "ab/c",
			"ab\\/\\//c" to "ab\\/\\//c",
			"a/" to "a/",
			"a//" to "a/",
			"/a" to "/a",
			"//a" to "/a",
			"/" to "/",
			"\\/" to "\\/",
			"\\" to "\\",
			"\\\\" to "\\\\"
		).forEach { (input, expected) ->
			Assertions.assertEquals(expected, input.removeDuplicatedSlashes())
		}
	}
}